src/main/java/unstoppableResolution/CNS.java

```diff
@@ -0,0 +1,109 @@
+
+package unstoppableResolution;
+
+import java.math.BigInteger;
+import java.net.UnknownHostException;
+
+import org.web3j.protocol.Web3j;
+import org.web3j.tx.gas.DefaultGasProvider;
+
+import unstoppableResolution.contracts.generated.Cryptoregistry;
+import unstoppableResolution.contracts.generated.Cryptoresolver;
+
+public class CNS extends NamingService {
+  final static String registryAddress = "0xD1E5b0FF1287aA9f9A268759062E4Ab08b9Dacbe";
+  Cryptoregistry registryContract;
+
+  public CNS(Web3j web3, Boolean verbose) {
+    super("CNS", web3, verbose);
+    this.registryContract = this.buildRegistry();
+  }
+
+  public Boolean isSupported(String domain) {
+    String[] split = domain.split("\\.");
+    return (split.length != 0 && split[split.length - 1].equals("crypto"));
+  }
+
+  public String addr(String domain, String ticker) throws NamingServiceException {
+    this.owner(domain); // If owner is not found NSExceptionCode.UnregisteredDomain is raised
+    String key = "crypto." + ticker.toUpperCase() + ".address";
+    String address = this.resolveKey(key, domain);
+    if (this.isNull(address))
+      throw new NamingServiceException(NSExceptionCode.UnknownCurrency, domain);
+    return address;
+  }
+
+  public String ipfsHash(String domain) throws NamingServiceException {
+    String key = "ipfs.html.value";
+    return this.resolveKey(key, domain);
+  }
+
+  public String email(String domain) throws NamingServiceException {
+    String key = "whois.email.value";
+    return this.resolveKey(key, domain);
+  }
+
+  public String owner(String domain) throws NamingServiceException {
+    try {
+      BigInteger tokenID = this.tokenID(domain);
+      String owner = this.owner(tokenID);
+      if (this.isNull(owner))
+        throw new NamingServiceException(NSExceptionCode.UnregisteredDomain, domain);
+      return owner;
+    } catch (NamingServiceException e) {
+      throw e;
+    } catch (UnknownHostException e) {
+      throw new NamingServiceException(NSExceptionCode.BlockchainIsDown, "CNS");
+    } catch (Exception e) {
+      throw new NamingServiceException(NSExceptionCode.UnknownError, e.getStackTrace().toString());
```

Comment:
In java there is built-in way to chain exceptions. You may specify the cause of exception as other exception: https://www.baeldung.com/java-chained-exceptions#throwable-class

Please use that instead of manually passing the stack trace.